### DISCUSSION TITLE
How to show a 3-D CAD file ( .step or .stl or .ply ) in egui ?

### DISCUSSION THREAD

**Discussion by Shrav108** [emilk/egui#5051] [Q&A]
I want to show cad file in my application at run time. How to show a .step or .stl or .ply or .obj files in egui ?

**Comment by enomado** [comment]
First of all egui runs in 2d space. So AFAIK you need to use 3d engine like three-d, or bevy. three-d had some troubles running inside egui(mostly in input events). So you need full 3d rendering with lighting, scene, etc.

Then you need to bake your cad model into model that 3d engine can read. 
three-d has https://github.com/asny/three-d-asset

Bevy uses something like https://docs.rs/bevy/latest/bevy/gltf/index.html

**Reply by atoktoto** [reply]
I'm also interested in this. Is there a tutorial / example of nesting bevy in egui window somewhere?

**Reply by Shrav108** [reply]
It seems there are two ways to do this : 
1. with ```three-d```
2. with ```bevy```

-> With ```bevy``` reading .stl files files is easy with ```bevy_stl```, but embedding a ```bevy``` app (reading a .stl file) into a ```egui``` app is a big question.

-> With ```three-d```, they gave me this https://github.com/emilk/egui/commit/636a39cbe122ce91ea554920ee7aa4475d8f2de7 link which embeds ```three-d``` app into ```egui```, but I am running into trouble with ```glow``` crate. It seems that different crates are using different versions of ```glow```. The app is not compiling for me. Please check it and if successful please let us know.

**Reply by enomado** [reply]
They try to sync use same version as egui https://github.com/asny/three-d/pull/466#issuecomment-2298513206

my `cargo tree --invert glow` says


three-d v0.17.0
egui_glow v0.28.1
glow v0.13.1

and there is no other versions of glow

**Reply by enomado** [reply]
[Shrav108](https://github.com/Shrav108)

I made quick example from my old code https://github.com/enomado/eframe_three_d_sample
It should just run

Wish I had time to investigate controls...

**Comment by YgorSouza** [comment]
I don't know if that helps, as I've never used these crates, but [transform-gizmo](https://github.com/urholaukkarinen/transform-gizmo) lets you manipulate 3D objects inside egui, and [stl_io](https://crates.io/crates/stl_io) lets you load a .stl file into a list of triangles. It's what bevy_stl uses.

**Reply by enomado** [reply]
I run [transform-gizmo](https://github.com/urholaukkarinen/transform-gizmo)  example.
its so fun. they project 3d on 2d space. this looks so strange
<img width="199" alt="image" src="https://github.com/user-attachments/assets/b8a54f78-efae-44af-8ef8-326a51467df6">
